### PR TITLE
Fix: User specified content type is ignored

### PR DESCRIPTION
**Pull request by ajaybhargavb**
#2947 

@rynowak

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewExecutor.cs [hunk 1]
```diff
@@ -36,7 +36,20 @@ public static class ViewExecutor
         {
             var response = actionContext.HttpContext.Response;
 
-            contentType = contentType ?? DefaultContentType;
+            if (contentType == null)
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewExecutor.cs]: Not sure if we want to consider other action results which have similar issue (ex: https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Core/ContentResult.cs#L40)

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewExecutor.cs]: Good point. I'll update it and check for other results with the same issue.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewExecutor.cs [hunk 2]
```diff
@@ -36,7 +36,20 @@ public static class ViewExecutor
...
+            {
+                contentType = DefaultContentType;
+
+                if (!string.IsNullOrEmpty(response.ContentType))
+                {
+                    MediaTypeHeaderValue parsedContentType;
+                    if (MediaTypeHeaderValue.TryParse(response.ContentType, out parsedContentType))
+                    {
+                        contentType = parsedContentType;
+                    }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewExecutor.cs]: Why do we need to do this? Why does MVC care if you set an invalid content type?

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewExecutor.cs]: Hmm. Just curious, we already check if the encoding is set in a user supplied content type. Why not check for invalid ones?

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewExecutor.cs]: The WebAPI design for all of this was to make the header classes unable to accept invalid header values.

It was limiting in some scenarios where users had to interop with non-compliant other code - that's part of the reason why these are defined as strings on the http abstractions classes.

I think it's fine to add an encoding if it's valid... but we shouldn't mess with what the user set if it's invalid. That masks bugs.

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewExecutor.cs]: Got it. Will update.

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewExecutor.cs]: We still need to parse it. When the user specifies the encoding along with the content type, it needs to be used by the writer here. https://github.com/aspnet/Mvc/pull/3009/files#diff-62f5af455fe6d271af2d9522b4841df9L49

**Comment by ajaybhargavb** [comment]
Updated.

File: src/Microsoft.AspNet.Mvc.Core/ContentResult.cs [hunk 3]
```diff
@@ -34,31 +34,16 @@ public class ContentResult : ActionResult
...
+            if (contentTypeHeader != null && contentTypeHeader.Encoding == null)
+            {
+                // Do not modify the user supplied content type, so copy it instead
+                contentTypeHeader = contentTypeHeader.Copy();
+                contentTypeHeader.Encoding = Encoding.UTF8;
             }
-
-            response.ContentType = contentTypeHeader.ToString();
+            
+            response.ContentType = contentTypeHeader?.ToString() ?? response.ContentType ?? DefaultContentType.ToString();
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ContentResult.cs]: long line

**Comment by rynowak** [comment]
:shipit:

**Pull request merged**